source/distributed/fullydistributed_tria.cc

```diff
@@ -0,0 +1,562 @@
+// Copyright (C) 2008 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#include <deal.II/base/logstream.h>
+#include <deal.II/base/memory_consumption.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/distributed/fullydistributed_tria.h>
+#include <deal.II/distributed/p4est_wrappers.h>
+
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/sparsity_tools.h>
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace fullydistributed
+  {
+    namespace internal
+    {
+      template <int dim>
+      bool
+      parent(typename CellId::binary_type c, typename CellId::binary_type p)
+      {
+        if (c[0] != p[0])
+          return false; // has same coarse cell
+        if ((c[1] >> 2) != ((p[1] >> 2) + 1))
+          return false; // level has one difference
+
+
+        const unsigned int n_child_indices = p[1] >> 2;
+        const unsigned int children_per_value =
+          sizeof(CellId::binary_type::value_type) * 8 / dim;
+        unsigned int child_level  = 0;
+        unsigned int binary_entry = 2;
+
+        // Loop until all child indices have been written
+        while (child_level < n_child_indices)
+          {
+            Assert(binary_entry < c.size(), ExcInternalError());
+
+            for (unsigned int j = 0; j < children_per_value; ++j)
+              {
+                if ((((c[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)) !=
+                    (((p[binary_entry] >> (j * dim))) &
+                     (GeometryInfo<dim>::max_children_per_cell - 1)))
+                  return false;
+
+
+                ++child_level;
+                if (child_level == n_child_indices)
+                  break;
+              }
+
+            ++binary_entry;
+          }
+        return true;
+      }
+    } // namespace internal
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::reinit(
+      ConstructionData<dim, spacedim> &construction_data)
+    {
+      // check if there are locally relevant coarse-grid cells
+      if (construction_data.cells.size() == 0) // no
+        {
+          // 1) create a dummy hypercube
+          Point<spacedim> p1, p2;
```

Comment:
why do you need this? otherwise the triangulation will return ``empty()==true`` which breaks other code?

Reply:
The library will break if I remove it because it implicitly assumes that there is at least one cell on each process (which is naturally the case for `p:d:t`, since the coarse grid is duplicated; even if number of procs is bigger than number of active cells).

Reply:
If a user provides an invalid input argument (such as an empty `construction_data`), then catch that via an assertion, rather than working around it.

Reply:
Or are you saying that it's a valid thing to do because one might not have a locally owned cell on all processors? I'd much rather us fix the breakages that result, rather than work around things by creating "artificial" cells that don't actually match anything elsewhere -- you're going to get into all sorts of problems downstream with your approach because you now have these kinds of cells. For example, you can now no longer rely on the fact that artificial cells are real cells with real coordinates at some level of the hierarchy. In other words, you can't trust anything any more about the geometry of the mesh.

Reply:
That a program aborts if one process does not have any cells is in my opinion too restrictive: given adaptivity, you might start with a very coarse mesh and then refine repeatedly (in this case the program would crash at the beginning).

I agree with you fear regarding the *artificial* information about the geometry. The motivation for doing the way I did it was the assumption that you would/should not access artificial cells.

I will have look into an alternative approach and I will write you. My fear is that both `create_triangulation` and `update_cache` (of `dealii::Triangulation`) will have issues with a mesh definition without cells. Would it be a solution to enable `dealii::Triangulation` to have zero cells? If yes, I would open a new PR and write a test.

Reply:
@bangerth How do we want to proceed with this issue? Do we want in the case that there are processes without locally relevant cells that these processes have a single artificial cell (as it is implemented right now) or that they have no cells.

Both have pros and cons. In the case of no cells, (at least) a couple of asserts have to disabled e.g. in `DoFHandlerPolicy` so that they can handle triangulations without levels. I must admit I do not have a good feeling about removing those asserts. Furthermore there are - as far I as I see and first tests show - parts in the library which assume that there is at least one coarse cell...

Reply:
@bangerth Would it be a solution to have a single artificial cell. But instead of creating it with the help of `GridGenerator::hypercube`, I use any global coarse-grid cell and set it artificial.